### DISCUSSION TITLE
setQueryData should provide an option to NOT reset isInvalidated to false

### DISCUSSION THREAD

**Discussion by kostia1st** [TanStack/query#5393] [Ideas]
`setQueryData` should provide an option to NOT reset `isInvalidated` to `false`.

The use case is kind of common.

I have a typical application that fetches some data, and sometimes mutates the data.
Once mutation is done on the server, sometimes I want the client data to reload (invalidate), and sometimes I just update it locally, and that's it. 

The problem is when I need a combination of both.
And here I start to depend on the order of calling `setQueryData` and `invalidateQueries`. Because if I first `invalidateQueries` to designate some necessary update (including _inactive_ queries, this is important), and then run `setQueryData` which for instance makes some kind of optimistic update, or maybe does not update anything important, but still assigns a formally new value to some old key (albeit being basically the same) - the `isInvalidated` flag of that _inactive_ query gets reset to `false`, and no update will be happening on activation of said (eternally fresh in my setup btw) query.

Currently I just work it around with the use of `getQueryState` before `setQueryData` and then `invalidateQueries` with `exact: true` afterwards to bring `isInvalidated` back to `true` if necessary.

But would love the ability to supply some extra option into that `setQueryData`'s 3rd parameter to prohibit resetting `isInvalidated`

**Comment by TkDodo** [comment]
`setQueryData` puts data in the cache. For the cache, it doesn't matter where data comes from. Could be from `setQueryDat`, but could also be from `prefetchQuery`, or from executing the `queryFn`, or via hydration, or ...

If data gets in the cache, it's considered as accurate. If your data is "older", you can pass `dataUpdatedAt` to `setQueryData` (last param), so that we know when the data was created. Hydration does this, as well as persisters.

This will make sure `staleTime` will be respected, but it will not ensure that data that was manually invalidated will then also be marked as "invalid".

**Reply by kostia1st** [reply]
Well, my approach to state management is that `staleTime` is set to `Infinity`, so `updatedAt` does not matter and does not change anything (I checked before posting this issue). I understand the motivation of why `setQueryData` does what it does by default. I'm just asking for a bit more flexibility which results in more convenience of use, that's it. Luckily workaround is there, although I'm not sure if it does influence CPU performance or not.
